python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1107 @@
+
+    Parameters
+    ----------
+    pull_func : func
+        UDF pull handler
+    """
+    def __init__(self, pull_func):
+        self.pull_func = pull_func
+
+    def __getstate__(self):
+        return self.pull_func
+
+    def __setstate__(self, state):
+        self.pull_func = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        kv.pull_handler = self.pull_func
+        res = RegisterPullHandlerResponse(kv.server_id)
+        return res
+
+REGISTER_PUSH = 901236
+
+class RegisterPushHandlerResponse(rpc.Response):
+    """Send a confirmation signal (just a server ID) of
+    RegisterPushHandler to client.
+
+    Parameters
+    ----------
+    server_id : int
+        ID of current server
+    """
+    def __init__(self, server_id):
+        self.server_id = server_id
+
+    def __getstate__(self):
+        return self.server_id
+
+    def __setstate__(self, state):
+        self.server_id = state
+
+class RegisterPushHandlerRequest(rpc.Request):
+    """Send a UDF to register Push handler on server.
+
+    Parameters
+    ----------
+    push_func : func
+        UDF push handler
+    """
+    def __init__(self, push_func):
+        self.push_func = push_func
+
+    def __getstate__(self):
+        return self.push_func
+
+    def __setstate__(self, state):
+        self.push_func = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        kv.push_handler = self.push_func
+        res = RegisterPushHandlerResponse(kv.server_id)
+        return res
+
+GET_SHARED = 901237
+
+class GetSharedDataResponse(rpc.Response):
+    """Send meta data of shared-tensor back to client.
+
+    Parameters
+    ----------
+    meta : dict
+        a dict of meta, e.g,
+
+        {'data_0' : (shape, dtype, policy_str),
+         'data_1' : (shape, dtype, policy_str)}
+    """
+    def __init__(self, meta):
+        self.meta = meta
+
+    def __getstate__(self):
+        return self.meta
+
+    def __setstate__(self, state):
+        self.meta = state
+
+class GetSharedDataRequest(rpc.Request):
+    """Send a signal (just a client ID) to get the
+    meta data of shared-tensor from server.
+
+    Parameters
+    ----------
+    client_id : int
+        ID of current client
+    """
+    def __init__(self, client_id):
+        self.client_id = client_id
+
+    def __getstate__(self):
+        return self.client_id
+
+    def __setstate__(self, state):
+        self.client_id = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        meta = {}
+        for name, data in kv.data_store.items():
+            meta[name] = (F.shape(data), 
+                          get_type_str(F.dtype(data)), 
+                          kv.part_policy[name].policy_str())
+        res = GetSharedDataResponse(meta)
+        return res
+
+GET_FULL_SHAPE = 901238
+
+class GetFullShapeResponse(rpc.Response):
```

Comment:
only shape? no dtype?

Reply:
We can get dtype in the get_shared_data() api.